Memory Cards: Use file memory mapping instead of explicit file I/O

F0bes opened the pull request:
### Description of Changes
Instead of freads for every memory card read, and fwrite for every memory card write, map the entire memory card into a memory mapping and let the OS manage caching / paging and be a little more efficient.

I've also implemented a separate thread that will asynchronously flush the memory map and the file caches. This triggers when our heuristic to detect memory card saves determines that the memory card has been saved to. It doesn't flush on every game write to the memory card, since a save can take many writes to the memory card. It all depends on the game. With this implementation, we benefit from the speed improvement from memory mapped file access and most importantly **reduces the time window for external events such as BSoD, AC power less, etc to result in corrupted memory card files.**
 
### Rationale behind Changes
This completely eliminated a bottleneck I found on the Persona 3 memory card menu. Overall memory card reads and block clearing should be faster with less EE thread usage.

As stated above, the time window for your memory card to be corrupted due to power loss or other external events becomes smaller. Beforehand, we had no way of knowing whether the memory card still had dirty pages in memory or not.

This should hopefully address #12433

### Suggested Testing Steps

## This is experimental and with all memory card code changes, scary! Please do not try this on memory cards you don't want to possibly lose.

To test you just have see if memory card stuff works. Saving / formatting / reading.

F0bes committed:
Memory Cards: Use file memory mapping instead of explicit file I/O

kamfretoz reviewed (approved):
Works good on both linux and mac

F0bes committed:
Config: Delete an implicitly deleted operator overload.

This fixes the two terabytes of logging we get to search through when we want to find something in a macOS actions run.

kamfretoz reviewed (changes requested):
Formatting an **unformatted folder memcard** causes PCSX2 to crash on linux with this backtrace:
```
*************** Unhandled SIGSEGV at 0x60f19fbec15e ***************
  0x0060f1a03e07f9 LogCallstack [/home/runner/work/pcsx2/pcsx2/common/CrashHandler.cpp:284]
  0x0060f1a03e06e3 CrashSignalHandler [/home/runner/work/pcsx2/pcsx2/common/CrashHandler.cpp:315]
  0x0060f1a0400853 SignalHandler [/home/runner/work/pcsx2/pcsx2/common/Linux/LnxHostSys.cpp:362]
  0x00754dd22421ff
  0x0060f19fbec15e ReadFromFile [/home/runner/work/pcsx2/pcsx2/pcsx2/SIO/Memcard/MemoryCardFolder.cpp:909]
  0x0060f19fbec3a6 ReadDataWithoutCache [/home/runner/work/pcsx2/pcsx2/pcsx2/SIO/Memcard/MemoryCardFolder.cpp:1014]
  0x0060f19fbec3a6 Read [/home/runner/work/pcsx2/pcsx2/pcsx2/SIO/Memcard/MemoryCardFolder.cpp:972]
  0x0060f19fbfced5 Read [/home/runner/work/pcsx2/pcsx2/pcsx2/SIO/Memcard/MemoryCardFolder.cpp:2332]
  0x0060f19fbfced5 FileMcd_Read [/home/runner/work/pcsx2/pcsx2/pcsx2/SIO/Memcard/MemoryCardFile.cpp:768]
  0x0060f19fbfced5 Read [/home/runner/work/pcsx2/pcsx2/pcsx2/SIO/Sio.h:51]
  0x0060f19fbfced5 ReadData [/home/runner/work/pcsx2/pcsx2/pcsx2/SIO/Memcard/MemoryCardProtocol.cpp:246]
  0x0060f19fbdc643 Memcard [/home/runner/work/pcsx2/pcsx2/pcsx2/SIO/Sio2.cpp:301]
  0x0060f19fbdc643 Write [/home/runner/work/pcsx2/pcsx2/pcsx2/SIO/Sio2.cpp:441]
  0x0060f19fb80505 psxDma11 [/home/runner/work/pcsx2/pcsx2/pcsx2/IopDma.cpp:241]
  0x0060f1ddbce010
*******************************************************************
Aborting application.
fish: Job 1, './PCSX2-linux-Qt-x64-appimage…' terminated by signal SIGABRT (Abort)
```

F0bes committed:
Darwin: Don't pass MAP_ANON to mmap when file mapping and mmap ret -1.

This appeared to work under x86, but on ARM this (rightfully) fails.

F0bes commented:
I don't actually modify how folder memory cards work. Can you confirm if this issue was introduced by this PR?

kamfretoz commented:
I confirm that this issue was not introduced by this PR as it was also crashing on master.
